PR: Consistently handle PYTHONPATH and add Import feature to PYTHONPATH Manager

mrclary opened the pull request:
<!--- Make sure to read the Contributing Guidelines:                   --->
<!--- https://github.com/spyder-ide/spyder/blob/master/CONTRIBUTING.md --->
<!--- and follow PEP 8, PEP 257 and Spyder's code style:               --->
<!--- https://github.com/spyder-ide/spyder/wiki/Dev:-Coding-Style      --->

## Description of Changes


Explicitly excluded system `PYTHONPATH` paths from Editor completion paths, IPython Console `sys.path`, and IPython Console `os.environ['PYTHONPATH']`.
Added feature to import system `PYTHONPATH` paths into PYTHONPATH Manager.

The "Syncronize..." feature (Windows only) is changed to "Export" to better complement "Import" and comport with what it actually does. This remains a Windows only feature for the time being.

Resolved an issue where a path was added after canceling "Select directory" dialog.

![ppm](https://user-images.githubusercontent.com/9618975/165896851-1bb4987a-08b3-4308-b3cd-e64cb7ab1bf9.gif)

### Issue(s) Resolved

<!--- List the issue(s) below, in the form "Fixes #1234"; one per line --->

Fixes #17511 


### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

<!--- TYPE YOUR USER/NAME AFTER THE FOLLOWING: --->
I certify the above statement is true and correct:

<!--- Thanks for your help making Spyder better for everyone! --->

pep8speaks commented:
Hello @mrclary! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:









* In the file [`spyder/utils/icon_manager.py`](https://github.com/spyder-ide/spyder/blob/11dec2a530ca3a945d41d6dba7117d31204e611d/spyder/utils/icon_manager.py):

> [Line 212:80](https://github.com/spyder-ide/spyder/blob/11dec2a530ca3a945d41d6dba7117d31204e611d/spyder/utils/icon_manager.py#L212): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (88 > 79 characters)

mrclary commented:
We had been removing `PYTHONPATH` from `env_vars` and using `SPY_PYTHONPATH` instead. This is only used to set the `sys.path` paths in the IPython Console at launch. Eliminating `SPY_PYTHONPATH` in favor of `PYTHONPATH` will produce the same result while also providing `os.environ['PYTHONPATH']` in the IPython Console and obviate `set_spyder_pythonpath` in `spyder-kernels`.

A companion PR spyder-ide/spyder-kernels#378 accommodates this change and updates IPython Console's `os.environ['PYTHONPATH']` along with the `sys.path` update.

ccordoba12 commented:
Thanks for your work on this @mrclary! I'm leaving this for 5.3.1 because this is an important change and we need to carefully discuss about it.

mrclary commented on spyder/app/tests/test_mainwindow.py:
```diff
@@ -3470,31 +3468,24 @@ def test_path_manager_updates_clients(qtbot, main_window, tmpdir):
-                syspath = shell.execute(cmd)
-
-            # `shell.executed` signal was not working so we use waitUntil
-            qtbot.waitUntil(lambda: test_folder in control.toPlainText(),
-                            timeout=SHELL_TIMEOUT)
-            assert test_folder in control.toPlainText()
```
@goanpeca @dalthviz , why should `qtbot` wait for `test_folder in control.toPlainText()` to be True? This is the assertion condition, so if the shell failed to update, `qtbot` will unnecessarily wait for 20-30 seconds before moving forward to the assertion statement. If the intent is to ensure that the text is written to the console before checking the assertion, then isn't this satisfied by the previous `qtbot.waitSignal(shell.executed)`? Perhaps a `SHELL_TIMEOUT` added here as well?

ccordoba12 commented on spyder/app/tests/test_mainwindow.py:
```diff
@@ -3453,8 +3453,6 @@ def test_runcell_cache(main_window, qtbot, debug):
 @pytest.mark.slow
 @flaky(max_runs=3)
-@pytest.mark.skipif(not sys.platform.startswith('linux'),
-                    reason="Works reliably on Linux")
```
To be honest, I don't remember. Sometimes our test suite starts to fail at random and I have to add skip's here and there to avoid that.

ccordoba12 commented:
@mrclary, we discussed your PR with @CAM-Gerlach and @isabela-pf this week and we consider it requires some UI/UX improvements, so that users better understand how the PYTHONPATH manager works with the improvements you propose.

@isabela-pf will be posting more details about it soon.

mrclary commented:
> @mrclary, we discussed your PR with @CAM-Gerlach and @isabela-pf this week and we consider it requires some UI/UX improvements, so that users better understand how the PYTHONPATH manager works with the improvements you propose.
> 
> @isabela-pf will be posting more details about it soon.

Thanks, @ccordoba12 @CAM-Gerlach and @isabela-pf .

Another thing to discuss is whether we should intercept `PYTHONPATH` in Spyder at launch. In this PR I have only ensured that `PYTHONPATH` is handled consistently between platforms, launch mechanisms, and completions/IPython Console. @ccordoba12 implemented `SPY_PYTHONPATH` to intercept PYTHONPATH Manager paths and re-inject them into `sys.path` in IPython Console without crashing the kernel if the user shadowed standard libraries. This was a good idea and I'm wondering if we shouldn't implement something similar for Spyder at launch. Spyder does not need the user's system `PYTHONPATH` for anything, but it does get injected into Spyder's runtime `sys.path`, so if it does shadow standard libraries it could cause problems for Spyder. Although I don't think we've seen any issues posted, so maybe it's nothing to be concerned about, it seems a relatively cheap thing to do to ensure Spyder is robust against user settings.

mrclary commented:
@isabela-pf, did you have some feedback on this PR?

mrclary commented:
@isabela-pf, thanks for responding. 

@spyder-ide/core-developers 
> * Since (from my understanding) they behave differently, I think there should be some kind of marker to label what is manually added versus what is added with the new feature. (This might be helpful to know after like for exporting or to remove things.)
> * Reorder the buttons to group them by use

I not sure what the above means; perhaps we can discuss at our meeting on Thursday.

> * Consider having more descriptive button names, like "Import from PYTHONPATH." Even though it's longer, the clarity might be better.

Is the hover text insufficient?
<img width="326" alt="Screen Shot 2022-04-26 at 12 43 27 PM" src="https://user-images.githubusercontent.com/9618975/165379793-5b215546-9f83-4447-9ad6-9c375221132a.png">

> * @CAM-Gerlach recommended a multicolumn approach to paths by how they are brought in. this seems like a lot.

I don't see any purpose for designating the origin of paths in the PYTHONPATH Manager. Whether they are imported or manually entered, it is just a simple list that the user may do with what they wish (remove, reorder, disable).

> * Also, should this be a new section in Preferences since it is a global Spyder change?

I don't know what you mean by "global" Spyder change, nevertheless, I don't think there is a need for a section in Preferences related to anything in this PR.

mrclary commented:
Okay, per conversation with @CAM-Gerlach today, I will tidy up the UI
- Add explanation at the top of the PYTHONPATH Manager dialog explaining what `PYTHONPATH` does and what exactly Spyder does with it, so the user know what to expect.
- Ensure that the Add path, Remove path, Import, and Export buttons are grouped together on the left of the dialog.
- Ensure proper tool tips for the Import and Export buttons

If more clarifications are needed or anyone has further questions or concerns, please reach out.

isabela-pf commented:
I'm glad @CAM-Gerlach was able to clarify! I don't want to block this PR with my limited time, so please don't feel the need to wait for me.

mrclary commented:
@ccordoba12 @CAM-Gerlach, I think I've addressed the final issues. If there are any revisions, please let me know. I've updated the PR description as well.

mrclary commented:
@ccordoba12 @CAM-Gerlach @dalthviz Are we waiting on anything for this PR?

CAM-Gerlach commented on spyder/widgets/pathmanager.py:
```diff
@@ -71,9 +71,17 @@ def __init__(self, parent, path=None, read_only_path=None,
 
         # Layouts
+        description = QLabel(
+            _("Only paths listed here will be passed to IPython consoles and "
+              "the language server<br>as additional Python search paths."
+              "<br>"
+              "Paths in your system <b>PYTHONPATH</b> environment variable "
+              "can be imported to this list.")
```
I tried to make the wording a little clearer and more understandable, though I'm still not 100% happy with it. I also elided the hard breaks for now, since they won't line up the same as before, and may not produce the desired results at different window widths, UI scaling, font sizes and translations; if you want the second break, it should generally be a full double paragraph `<br>`.

```suggestion
            _("The paths listed below will be passed to IPython consoles and "
              "the language server as additional locations to search for "
              "Python modules. "
              "Any paths in your system <b>PYTHONPATH</b> environment "
              "variable can be imported here if you'd like to use them.")
```

CAM-Gerlach commented on spyder/widgets/pathmanager.py:
```diff
@@ -202,15 +214,35 @@ def setup(self):
-            _("Synchronize"),
-            _("This will synchronize Spyder's path list with "
+            _("Export"),
+            _("This will export Spyder's path list to "
```
```suggestion
            _("This will export Spyder's path list to the "
```

Missing word

CAM-Gerlach commented:
Sorry for the delay; I've been at PyCon US for the past >week, just got back yesterday and have been trying to catch up. This looks pretty good, aside from a few textual tweaks :+1:

ccordoba12 reviewed (commented):
Thanks @mrclary for your patience with this one! I think it's shaping very nicely.

mrclary commented on spyder/widgets/pathmanager.py:
```diff
@@ -71,9 +71,18 @@ def __init__(self, parent, path=None, read_only_path=None,
         self.resize(500, 300)
-        self.sync_button.setVisible(os.name == 'nt' and sync)
+        self.import_button.setVisible(sync)
+        # self.export_button.setVisible(os.name == 'nt' and sync)
```
```suggestion
        self.export_button.setVisible(os.name == 'nt' and sync)
```

mrclary commented on spyder/widgets/pathmanager.py:
```diff
@@ -202,20 +215,40 @@ def setup(self):
+    def export_pythonpath(self):
         """
-        Synchronize Spyder's path list with PYTHONPATH environment variable
-        Only apply to: current user, on Windows platforms.
```
No, sorry; thanks for catching that. See my comment for line 75.

mrclary commented on spyder/plugins/completion/providers/languageserver/provider.py:
```diff
@@ -799,15 +800,15 @@ def generate_python_config(self):
                                         section='main_interpreter')
-            env_vars = os.environ.copy()
-            # external interpreter should not use internal PYTHONPATH
-            env_vars.pop('PYTHONPATH', None)
```
Oops, I missed that. You are correct, Jedi does not need `PYTHONPATH` and should not have it set. We pass all the extra paths via `extra_paths`

mrclary commented on spyder/plugins/completion/providers/languageserver/provider.py:
```diff
@@ -799,15 +800,15 @@ def generate_python_config(self):
         }
 
         # Jedi configuration
+        env_vars = os.environ.copy()  # Ensure env is indepependent of PyLSP's
```
```suggestion
        env_vars = os.environ.copy()  # Ensure env is indepependent of PyLSP's
        env_vars.pop('PYTHONPATH', None)
```

mrclary commented on spyder/widgets/pathmanager.py:
```diff
@@ -202,20 +215,40 @@ def setup(self):
 
     @Slot()
-    def synchronize(self):
+    def import_pythonpath(self):
```
Okay, I'll implement.

CAM-Gerlach replied:
> This action cannot be easily reversed. So, if users have a PYTHONPATH with say 20 paths and click on this but they really don't want to use them in Spyder, they will be forced to remove all those paths from our manager. Therefore, I think it's better to ask them if they really want to proceed before doing that.

Allowing multiple selection (Shift, Ctrl, etc) selection would greatly improve usability here, but presumably the button actions would need to be updated to handle that; I'm not sure how difficult that would be to do.

ccordoba12 commented on spyder/widgets/pathmanager.py:
```diff
@@ -202,20 +215,52 @@ def setup(self):
+        answer = QMessageBox.question(
+            self,
+            _("Import"),
+            _("Do you want to import the following paths into "
+              "PYTHONPATH Manager?"
```
```suggestion
            _("Do you want to import the contents of your <tt>PYTHONPATH</tt> "
              "environment variable into Spyder? The following paths will be "
              "imported:"
```

* I think we can be clearer and more explicit here.
* I like that users can check the paths that will be imported. Thanks for adding that.

ccordoba12 commented on spyder/widgets/pathmanager.py:
```diff
@@ -202,20 +215,52 @@ def setup(self):
+    def import_pythonpath(self):
+        """Import from PYTHONPATH environment variable"""
+        env_pypath = os.environ.get('PYTHONPATH', '')
+        if not env_pypath:
+            return
```
This is confusing because the button does nothing when clicked, which users could interpret as it being broken. So, in this case I suggest to add a `QMessageBox.information` saying that `PYTHONPATH` is empty, so there's nothing to import from it.

mrclary replied:
No problem, I'll add another modal window.

ccordoba12 reviewed (commented):
I think this is almost ready, thanks @mrclary for the updates.

ccordoba12 commented on spyder/widgets/pathmanager.py:
```diff
@@ -202,20 +215,58 @@ def setup(self):
+            QMessageBox.information(
+                self,
+                _("PYTHONPATH"),
+                _("<tt>PYTHONPATH</tt> is empty; nothing to import."),
```
```suggestion
                _("Your <tt>PYTHONPATH</tt> environment variable is empty, so "
                  "there is nothing to import."),
```

Sorry to be so picky about these messages, but I think the more explicit we are in them, the easier would be for users to understand what this is about in case they don't know it.

For example, by saying that `PYTHONPATH` is an env var, it'd be easier for them to google for it.

mrclary replied:
No problem.

ccordoba12 reviewed (commented):
@mrclary, besides my small suggestion below, I think we should leave the feature of adding paths to the front of `sys.path` for another PR.

That's because if we want to make that configurable (as we agreed on https://github.com/spyder-ide/spyder-kernels/pull/378), I think we need to change the `update_syspath` handler on Spyder-kernels. And that requires a new minor Spyder-kernels version (i.e. 2.4.0) and a corresponding minor Spyder version too (i.e. 5.4.0).

mrclary commented:
> @mrclary, besides my small suggestion below, I think we should leave the feature of adding paths to the front of `sys.path` for another PR.
> 
> That's because if we want to make that configurable (as we agreed on [spyder-ide/spyder-kernels#378](https://github.com/spyder-ide/spyder-kernels/pull/378)), I think we need to change the `update_syspath` handler on Spyder-kernels. And that requires a new minor Spyder-kernels version (i.e. 2.4.0) and a corresponding minor Spyder version too (i.e. 5.4.0).

I'll submit a (draft) PR for adding the feature.

mrclary commented:
/show binder

CAM-Gerlach commented:
> Additionally, it is not using the correct external-deps spyder-kernels.

Perhaps because it is running normally using the PyPI deps instead of installing the subrepos?

mrclary commented:
> > Additionally, it is not using the correct external-deps spyder-kernels.
> 
> Perhaps because it is running normally using the PyPI deps instead of installing the subrepos?

The behavior I see suggests that may be what is happening. But I don't know anything about the binder setup to say anything intelligent...

mrclary commented:
> > > Additionally, it is not using the correct external-deps spyder-kernels.
> > 
> > 
> > Perhaps because it is running normally using the PyPI deps instead of installing the subrepos?
> 
> The behavior I see suggests that may be what is happening. But I don't know anything about the binder setup to say anything intelligent...

Okay, so I forgot to add back the subrepo clone into this PR; that's why binder was mis behaving. I just tested it now and it is working as expected, as far as this PR is concerned. The jellyfish, pyls_spyder, and pylsp_black issue is still present.

CAM-Gerlach commented:
>  The jellyfish, pyls_spyder, and pylsp_black issue is still present.

Strange, since you added it to the binder environment.yaml at the same time as the conda, etc. requirements, and it indeed [appears to be present](https://github.com/mrclary/spyder/blob/pypath-manager-import/binder/environment.yml) on your branch. You could take a look with `conda list` on the Binder instance to see if it is present in the env

ccordoba12 commented:
> There appears to be some issues with binder.

We haven't updated the Binder deps in a long time (perhaps eight months). They are declared here (per this repo branch):

https://github.com/spyder-ide/binder-environments

Could you give us a hand with that?

> Okay, so I forgot to add back the subrepo clone into this PR

I think your changes in Spyder-kernels should be dropped from this PR and left for PR #18308. That way we could merge this one (which I think is ready) and you could keep working on solving issue #17066 there.

What do you think?

mrclary commented:
> > There appears to be some issues with binder.
> 
> We haven't updated the Binder deps in a long time (perhaps eight months). They are declared here (per this repo branch):
> 
> https://github.com/spyder-ide/binder-environments
> 
> Could you give us a hand with that?

No problem

> > Okay, so I forgot to add back the subrepo clone into this PR
> 
> I think your changes in Spyder-kernels should be dropped from this PR and left for PR #18308. That way we could merge this one (which I think is ready) and you could keep working on solving issue #17066 there.
> 
> What do you think?

I'm happy to do whatever you think is best. However, if spyder-ide/spyder-kernels#378 is not included with this PR, then paths in IPython Console `sys.path` will continue to be positioned differently depending on whether the user enabled/disabled PPM paths. If spyder-ide/spyder-kernels#378 is rolled into #18308, then it may be some time before this behavior is fixed since we'd have to wait until a minor version bump in both repos.

ccordoba12 commented:
> However, if https://github.com/spyder-ide/spyder-kernels/pull/378 is not included with this PR, then paths in IPython Console sys.path will continue to be positioned differently depending on whether the user enabled/disabled PPM paths

Ok, I see. What if you change https://github.com/spyder-ide/spyder-kernels/pull/378 so that paths are added to the bottom of `sys.path` for now? Then you could make that configurable on PR #18308.

CAM-Gerlach commented:
> We haven't updated the Binder deps in a long time (perhaps eight months). They are declared here (per this repo branch):

Is there a reason they are also declared under `binder/requirements.txt` in this repo, which has been kept up to date and is the location as documented in `requirements/conda.txt`? This was rather confusing to me, since I checked there as documented and they were all up to date.

mrclary commented:
> > However, if [spyder-ide/spyder-kernels#378](https://github.com/spyder-ide/spyder-kernels/pull/378) is not included with this PR, then paths in IPython Console sys.path will continue to be positioned differently depending on whether the user enabled/disabled PPM paths
> 
> Ok, I see. What if you change [spyder-ide/spyder-kernels#378](https://github.com/spyder-ide/spyder-kernels/pull/378) so that paths are added to the bottom of `sys.path` for now? Then you could make that configurable on PR #18308.

No problem.

ccordoba12 commented:
@mrclary, I can't test this PR because for me the `spyder-kernels` subrepo is not added to `sys.path` with your changes here. So, we need to merge PR #17408 first and then continue with this one.

mrclary commented:
> Hi @mrclary, I already tested this PR in mac using the installer and it is working as expected. My only comment is why the text in the buttons seems so small? I'm not sure if we have any dialog with the text so small.

I assume that you are referring to the question message box. I chose to try to distinguish the question from the list of paths using the `setInformativeText` method in order to make everything more readable. However, the following illustrate the different options.
- `setInformativeText`
  <img width="432" alt="Screen Shot 2022-06-20 at 2 28 38 PM" src="https://user-images.githubusercontent.com/9618975/174681187-26a16a3b-fc33-4ca1-91bf-348ae896c177.png">
- Combined with the question text
  <img width="421" alt="Screen Shot 2022-06-20 at 2 36 34 PM" src="https://user-images.githubusercontent.com/9618975/174681237-a72226b7-d16e-41f9-8b96-b1131f0fabbe.png">
- `setDetailedText`
  ![details](https://user-images.githubusercontent.com/9618975/174681599-04ae21be-435a-443a-8fe4-58fb1a04a544.gif)

Perhaps others have a different preference than mine?

In any case, it would be nice if the message box width adapted to greater-than-or-equal-to the width of the largest path string since distinguishing paths can be difficult if their text wraps. If anyone has a clever suggestion there, I would much appreciate it. Note that while the pats do not wrap for the `setInformativeText` case shown above, this is only circumstantial and may wrap depending on the length of the paths listed.

mrclary commented:
> @mrclary, I can't test this PR because for me the `spyder-kernels` subrepo is not added to `sys.path` with your changes here. So, we need to merge PR #17408 first and then continue with this one.

That's fine with me. Sorry that is working; I thought I sufficiently separated those two PRs, such that this one should not have affected bootstrapping behavior (sys.path manipulations under DEV should not have been touched). Maybe I missed something.

steff456 commented:
> Perhaps others have a different preference than mine?

I think of all the options you showed I prefer the third one, it will not overflow in case someone is importing a huge list. I know the wrap is inevitable, so that is not a huge blocker for me. Maybe adding something to recognize a new line may help (?) I have to check if we have something like that in the interface right now, but if not I think it is just fine as you have it.

mrclary commented:
I added a list marker for each path.

<img width="422" alt="Screen Shot 2022-06-20 at 6 32 49 PM" src="https://user-images.githubusercontent.com/9618975/174697963-ab241768-77c4-413b-bd6b-b41bd16261f1.png">

mrclary commented:
> @mrclary Can you change the UI text of the Show details button to `Show paths` (unless that is hardcoded in Qt)? This would make it more obvious to users what it does and where to actually view the paths that will be imported.

Unfortunately it appears to be hardcoded. I can locate the button and change the text, but the text updates upon clicking from "Show details..." to "Hide details...". I can change the initial text to "Show paths...", but cycling through the button restores "Show details...". I'd have to locate the method changing the text and patch it. Perhaps it would be easier to just create our own QWidget... Let me see what I can do.

mrclary commented:
@spyder-ide/core-developers how about the following?

![listwidget](https://user-images.githubusercontent.com/9618975/174850807-56a8beaa-e8af-4518-9f3b-7a6079965c4b.gif)

steff456 commented:
It looks better!

ccordoba12 commented:
Agreed, it looks much nicer.

ccordoba12 commented on spyder/widgets/pathmanager.py:
```diff
@@ -202,20 +215,68 @@ def setup(self):
+                _("PYTHONPATH"),
+                _("Your <tt>PYTHONPATH</tt> environment variable is empty, so "
+                  "there is nothing to import."),
+                QMessageBox.Ok)
```
```suggestion
                QMessageBox.Ok
            )
```

Style improvement.

ccordoba12 commented on spyder/widgets/pathmanager.py:
```diff
@@ -202,20 +215,68 @@ def setup(self):
+              "<br><br>"
+              "Do you want to clear the contents of PYTHONPATH before "
               "adding Spyder's path list?"),
             QMessageBox.Yes | QMessageBox.No | QMessageBox.Cancel)
```
```suggestion
            QMessageBox.Yes | QMessageBox.No | QMessageBox.Cancel
        )
```

Style improvement

ccordoba12 reviewed (commented):
Last minor details @mrclary, then this should be ready.

Also, since I just merged https://github.com/spyder-ide/spyder-kernels/pull/378, please resync the spyder-kernels subrepo with the instructions in our Contributing guide.

ccordoba12 reviewed (approved):
Looks good to me now, thanks @mrclary!

dalthviz reviewed (commented):
Thanks @mrclary ! Checking this locally on Windows seems like is working correctly although to import exported paths you need to restart Spyder and/or run it from a new console/cmd. Seems like the changes done to `PYTHONPATH`  are not being taken into account without that (I think that's something that was happening before this PR though, right?). So for example, if I use `Add a path`, use `Export`, remove the path from the manager with `Remove path` and then try to use `Import` I see the empty `PYTHONPATH` while in the consoles I can get the PYTHONPATH update:

![imagen](https://user-images.githubusercontent.com/16781833/175079966-4e387413-0383-4f54-a9b7-12daf2a38e32.png)

If that is expected then this LGTM :+1:

mrclary commented:
> Thanks @mrclary ! Checking this locally on Windows seems like is working correctly although to import exported paths you need to restart Spyder and/or run it from a new console/cmd. Seems like the changes done to `PYTHONPATH` are not being taken into account without that (I think that's something that was happening before this PR though, right?). So for example, if I use `Add a path`, use `Export`, remove the path from the manager with `Remove path` and then try to use `Import` I see the empty `PYTHONPATH` while in the consoles I can get the PYTHONPATH update:

The "import" feature currently uses the existing `os.environ['PYTHONPATH']` in Spyder's runtime environment which is established at launch. We could change this with one of the following options:
- Upon exporting PPM to the system PYTHONPATH (Windows only), we could also appropriately update Spyder's runtime `os.environ['PYTHONPATH']`. Spyder does not use this internally for any reason, so this should be safe.
- Rather than relying on `os.environ['PYTHONPATH']`, we could retrieve the system environment variable directly. This should be straightforward on Windows by querying the registry (I think this is already done for the purpose of exporting anyway). On unix, a login shell subprocess would be needed to extract the environment variable; this would require a little more work to setup properly as it depends on the user's default shell etc., but is certainly feasible.

The second option may be more robust and future-proof. For example, @CAM-Gerlach has alluded to pending updates to Python which may use new `-P` flags by default which will preclude use of `PYTHON*` environment variables. If/when this is the case, `os.environ['PYTHONPATH']` may no longer be available by default, so a different mechanism may be prudent.

> If that is expected then this LGTM 👍

Technically, it is expected. Nevertheless, it may not be desirable. I think I'll include a fix for this in this PR. Unless @ccordoba12, @dalthviz, or @CAM-Gerlach have anything other recommendations, I'll proceed with the second option.

mrclary commented:
> Sounds good to me 👍 Although, since the current behavior is expected maybe a new issue could be open to track the missing improvements (probably using the comment above [#17512 (comment)](https://github.com/spyder-ide/spyder/pull/17512#issuecomment-1163398698) as an starting point). In that way we could check a little bit more ideas and other missing things regarding the PYTHONPATH manager without that being a blocker to merge this

Good idea. I'll proceed with the second option in another pull request in order to allow more in-depth feedback without blocking this PR.

steff456 reviewed (approved):
Thanks for working on this! I think it is ready to merge 😬
